MoreLinq.NoConflictGenerator/Program.cs

```diff
@@ -0,0 +1,455 @@
+            var excludePredicate = PredicateFromPattern(excludePattern, false);
+
+            var thisAssemblyName = typeof(Program).GetTypeInfo().Assembly.GetName();
+
+            //
+            // Type abbreviations are used to abbreviate all generic type
+            // parameters into a letter from the alphabet. So for example, a
+            // method with generic type parameters `TSource` and `TResult` will
+            // become `a` and `b`. This is used later for sorting and stabilizes
+            // the sort irrespective of how the type parameters are named or
+            // renamed in the source.
+            //
+
+            var abbreviatedTypeNodes = Enumerable
+                .Range(0, 26)
+                .Select(a => (char) ('a' + a))
+                .Select(ch => new SimpleTypeKey(ch.ToString()))
+                .ToArray();
+
+            var qq =
+                from fp in Directory.EnumerateFiles(dir, "*.cs")
+                where !excludePredicate(fp) && includePredicate(fp)
+                orderby fp
+                from m in
+                    from cd in
+                        CSharpSyntaxTree
+                            .ParseText(File.ReadAllText(fp), CSharpParseOptions.Default.WithPreprocessorSymbols("MORELINQ"))
+                            .GetRoot()
+                            .SyntaxTree
+                            .GetCompilationUnitRoot()
+                            .DescendantNodes().OfType<ClassDeclarationSyntax>()
+                    where (string) cd.Identifier.Value == "MoreEnumerable"
+                    from md in cd.DescendantNodes().OfType<MethodDeclarationSyntax>()
+                    let mn = (string) md.Identifier.Value
+                    where md.ParameterList.Parameters.Count > 0
+                       && md.ParameterList.Parameters.First().Modifiers.Any(m => (string)m.Value == "this")
+                       && md.Modifiers.Any(m => (string)m.Value == "public")
+                       && md.AttributeLists.SelectMany(al => al.Attributes).All(a => a.Name.ToString() != "Obsolete")
+                    let typeParameterAbbreviationByName =
+                        md.TypeParameterList
+                         ?.Parameters
+                          .Select((e, i) => (Original: e.Identifier.ValueText, Alias: abbreviatedTypeNodes[i]))
+                          .ToDictionary(e => e.Original, e => e.Alias)
+                    select new
+                    {
+                        Syntax = md,
+                        Name = md.Identifier.ToString(),
+                        TypeParameterCount = md.TypeParameterList?.Parameters.Count ?? 0,
+                        TypeParameterAbbreviationByName = typeParameterAbbreviationByName,
+                        ParameterCount = md.ParameterList.Parameters.Count,
+                        SortableParameterTypes =
+                            from p in md.ParameterList.Parameters
```

Comment:
```c#
md.ParameterList
  .Parameters
  .Select(p => CreateTypeKey(p.Type,
                             n => typeParameterAbbreviationByName != null
                             && typeParameterAbbreviationByName.TryGetValue(n, out var a) ? a : null))
```

Not sure how it's any more concise? Is it because we're saving on `from` and `in`?

For me, this is not a question of style. The query syntax is almost always better because it's more scalable and enforces more rules at the language level that guide you correctly. I resort to method chaining syntax mostly when required. I say _almost_ and _mostly_ because there are very rare instances where it's worth breaking the rule but I don't see this as one of those instances.

Reply:
> The query syntax is almost always better because it's more scalable

Why do you say it's more scalable? Query syntax eventually compiles down to method syntax so I'm not following how it could be more scalable.

Reply:
I'm talking about a different metric for scale here, more from a code structure perspective than performance; possibly a poor choice of wording but sticking to it, try expressing a query with several nested loops/bindings using `from`…`in` and `SelectMany` as an example:

```c#
var q = Enumerable
    .Range(1, 3)
    .SelectMany(x => Enumerable.Range(1, x), (x, y) => new { X = x, Y = y })
    .SelectMany(e => Enumerable.Range(1, e.Y), (e, z) => new { e.X, e.Y, Z = z });
```

versus:

```c#
var q =
    from x in Enumerable.Range(1, 3)
    from y in Enumerable.Range(1, x)
    from z in Enumerable.Range(1, y)
    select new { X = x, Y = y, Z = z };
```

I hope you'll agree that the first one is no match for the second.